staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_cache.go

```diff
@@ -301,14 +331,56 @@ func (w *watchCache) processEvent(event watch.Event, resourceVersion uint64, upd
 
 // Assumes that lock is already held for write.
 func (w *watchCache) updateCache(event *watchCacheEvent) {
-	if w.endIndex == w.startIndex+w.capacity {
+	w.resizeCache(event.RecordTime)
+	if w.isCacheFullLocked() {
 		// Cache is full - remove the oldest element.
 		w.startIndex++
 	}
 	w.cache[w.endIndex%w.capacity] = event
 	w.endIndex++
 }
 
+// Use dynamic size cache. If all cached events occur inside eventFreshDuration, increase cache capacity by 2x.
+// Decrease cache capacity by 2x when recent half events occur outside eventFreshDuration.
+func (w *watchCache) resizeCache(eventTime time.Time) {
```

Comment:
nit: rename to resizeCacheLocked

Reply:
Done. As follow function which makes new event cache called resizeCacheLocked, rename old resizeCacheLocked to doCacheResizeLocked.